<a href="https://colab.research.google.com/github/Amrion/KIS/blob/main/%D0%A2%D0%B0%D1%80%D0%BD%D0%BE%D0%B2%D1%81%D0%BA%D0%B8%D0%B9_%D0%94_%D0%A0_%D0%98%D0%A35_83%D0%91_%D0%9B%D0%B0%D0%B1_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа 1. Формализация постановок задач исследования КИС

Тарновский Даниил, ИУ5-83Б, Вариант 19

## Ход лабораторной работы

Используем data.frame Computers:

In [ ]:
DataComputers<-read.csv("https://raw.githubusercontent.com/junaart/ForStudents/master/KIS/Lab_1/Computers.csv",sep=",")
View(DataComputers)

X    price speed hd   ram screen cd  multi premium ads trend
1     1   1499  25     80   4  14     no  no    yes     94  1    
2     2   1795  33     85   2  14     no  no    yes     94  1    
3     3   1595  25    170   4  15     no  no    yes     94  1    
4     4   1849  25    170   8  14     no  no    no      94  1    
5     5   3295  33    340  16  14     no  no    yes     94  1    
6     6   3695  66    340  16  14     no  no    yes     94  1    
7     7   1720  25    170   4  14     yes no    yes     94  1    
8     8   1995  50     85   2  14     no  no    yes     94  1    
9     9   2225  50    210   8  14     no  no    yes     94  1    
10   10   2575  50    210   4  15     no  no    yes     94  1    
11   11   2195  33    170   8  15     no  no    yes     94  1    
12   12   2605  66    210   8  14     no  no    yes     94  1    
13   13   2045  50    130   4  14     no  no    yes     94  1    
14   14   2295  25    245   8  14     no  no    yes     94  1    
15   15   2699  50    212   8  14     no  no    yes     94  1    
16   16   2225  50    130   4  14     no  no    yes     94  1    
17   17   1595  33     85   2  14     no  no    yes     94  1    
18   18   2325  33    210   4  15     no  no    yes     94  1    
19   19   2095  33    250   4  15     no  no    yes     94  1    
20   20   4395  66    452   8  14     no  no    yes     94  1    
21   21   1695  33    130   4  14     no  no    yes     94  1    
22   22   2795  66    130   4  14     no  no    yes     94  1    
23   23   2895  25    340  16  14     no  no    yes     94  1    
24   24   2875  50    210   4  17     no  no    yes     94  1    
25   25   4195  50    452   8  14     no  no    yes     94  1    
26   26   1290  33     80   2  14     no  no    yes     94  1    
27   27   1975  33    130   4  14     no  no    yes     94  1    
28   28   3995  33    452   8  14     no  no    yes     94  1    
29   29   3095  33    340  16  14     no  no    yes     94  1    
30   30   3244  66    245   8  14     no  no    yes     94  1    
⋮    ⋮    ⋮     ⋮     ⋮    ⋮   ⋮      ⋮   ⋮     ⋮       ⋮   ⋮    
6230 6230 2344   66    850 16  15     yes no    yes     39  35   
6231 6231 2890  100   1200 24  17     yes no    yes     39  35   
6232 6232 2390   66    850 16  17     yes no    yes     39  35   
6233 6233 1495   66    528  8  14     no  no    yes     39  35   
6234 6234 1844  100    528  8  14     no  no    yes     39  35   
6235 6235 1699   66    630  8  15     no  no    no      39  35   
6236 6236 3290   75   1600 32  17     yes no    yes     39  35   
6237 6237 2790   66   1200 24  17     yes no    yes     39  35   
6238 6238 2845  100   1600 24  15     yes no    yes     39  35   
6239 6239 1744   66    528  8  14     no  no    yes     39  35   
6240 6240 3095  100   1600 32  15     yes no    yes     39  35   
6241 6241 1595  100    528  8  14     no  no    yes     39  35   
6242 6242 2254  100    850 16  15     yes no    yes     39  35   
6243 6243 2723  100   1200 24  15     yes yes   yes     39  35   
6244 6244 2745   66   1600 24  15     yes no    yes     39  35   
6245 6245 3054   75   1600 32  15     yes no    yes     39  35   
6246 6246 2744   66   1200 24  15     yes no    yes     39  35   
6247 6247 1554   66    528  8  14     no  no    yes     39  35   
6248 6248 1654  100    528  8  14     no  no    yes     39  35   
6249 6249 2623   66   1200 24  15     yes yes   yes     39  35   
6250 6250 2095   66    850 16  15     yes no    yes     39  35   
6251 6251 2495   66   1200 24  15     yes no    yes     39  35   
6252 6252 2995   66   1600 32  15     yes no    yes     39  35   
6253 6253 1690   66   1200  8  14     no  no    yes     39  35   
6254 6254 2154   66    850 16  15     yes no    yes     39  35   
6255 6255 1690  100    528  8  15     no  no    yes     39  35   
6256 6256 2223   66    850 16  15     yes yes   yes     39  35   
6257 6257 2654  100   1200 24  15     yes no    yes     39  35   
6258 6258 2195  100    850 16  15     yes no 

Пояснения содержимого датафрейма:

A dataframe containing :

**price**

price in US dollars of 486 PCs

**speed**

clock speed in MHz

**hd**

size of hard drive in MB

**ram**

size of Ram in in MB

**screen**

size of screen in inches

**cd**

is a CD-ROM present ?

**multi**

is a multimedia kit (speakers, sound card) included ?

**premium**

is the manufacturer was a "premium" firm (IBM, COMPAQ) ?

**ads**

number of 486 price listings for each month

**trend**

time trend indicating month starting from January of 1993 to November of 1995.

Требуется развернуть КИС, включающую в себя:
- серверные станции $С_1$;
- рабочие станции аналитиков и менеджеров $C_2$;
- рабочие станции обслуживающего персонала $C_3$;
- станции сопряжения с видео, аудио оборудованием $C_4$.

Станции одного типа считать одинаковыми по всем параметрам.
Требования к каждому типу станций перечислены в таблицах ниже:

In [ ]:
set.seed(193) 
Count_low<-data.frame()
for (i in c(1:50)){
    Count_low<-rbind(Count_low, c(i,sample(c(3:6),1), sample(c(6:20),1), sample(c(1:20),1), sample(c(1:9),1)))}
colnames(Count_low)<-c("Вариант","Количество С1 >=","Количество С2 >=", "Количество С3 >=", "Количество С4 >=") 
View(Count_low)

Вариант Количество С1 >= Количество С2 >= Количество С3 >= Количество С4 >=
1   1      6                20               19               2               
2   2      4                18               13               2               
3   3      5                10               12               3               
4   4      4                12               19               3               
5   5      4                20               14               7               
6   6      3                10               15               8               
7   7      3                20               18               2               
8   8      5                20               10               2               
9   9      6                 9                4               9               
10 10      3                11                1               2               
11 11      6                20               13               7               
12 12      3                18               10               1               
13 13      3                13                1               5               
14 14      4                18                8               2               
15 15      3                 8                2               8               
16 16      5                 6               17               9               
17 17      3                15               12               2               
18 18      4                13                5               5               
19 19      6                20               11               7               
20 20      4                 9                1               6               
21 21      4                14               12               5               
22 22      6                15               17               6               
23 23      4                 9               11               4               
24 24      5                 8               14               5               
25 25      3                 9                4               8               
26 26      4                20               15               3               
27 27      6                 8                5               6               
28 28      3                13               18               9               
29 29      5                16                9               5               
30 30      3                15               12               9               
31 31      4                20               15               4               
32 32      4                14               14               9               
33 33      3                13               10               5               
34 34      4                 8               13               1               
35 35      4                18               13               3               
36 36      6                 9                2               7               
37 37      4                 9               14               8               
38 38      3                19               20               5               
39 39      5                10               20               4               
40 40      6                18               19               1               
41 41      3                12               12               3               
42 42      6                17                3               9               
43 43      3                14               18               7               
44 44      3                 9                2               2               
45 45      3                18               19               1               
46 46      3                12               12               7               
47 47      3                19               15               2               
48 48      3                10                6               2               
49 49      6                20                2               4               
50 50      4                16                9      

In [ ]:
Count_up<-data.frame()
for (i in c(1:50)){
    Count_up<-rbind(Count_up, c(i,Count_low[i,2]+sample(c(3:6),1), Count_low[i,3]+sample(c(6:20),1), 
                                          Count_low[i,4]+sample(c(1:20),1), Count_low[i,5]+sample(c(1:9),1)))}
colnames(Count_up)<-c("Вариант","Количество С1 <=","Количество С2 <=", "Количество С3 <=", "Количество С4 <=") 
View(Count_up)

Вариант Количество С1 <= Количество С2 <= Количество С3 <= Количество С4 <=
1   1      11               39               32                9              
2   2       9               38               22               10              
3   3      11               18               26                9              
4   4       7               19               28               11              
5   5       8               39               25               12              
6   6       7               27               20               10              
7   7       7               29               25                7              
8   8       9               40               28               10              
9   9      12               21               15               17              
10 10       8               31               12                5              
11 11      11               30               15               16              
12 12       8               30               26                4              
13 13       7               26                4               12              
14 14       8               32               24                6              
15 15       9               21               12               12              
16 16       8               22               37               15              
17 17       9               30               29                8              
18 18       7               31               13                7              
19 19      12               27               12               11              
20 20       9               23               21               13              
21 21       8               34               32                9              
22 22      11               24               37                9              
23 23       9               24               12                6              
24 24      10               25               32                7              
25 25       8               24               21               14              
26 26       8               30               29                6              
27 27      11               21               21               12              
28 28       6               25               34               18              
29 29       9               24               20                6              
30 30       7               26               29               16              
31 31      10               29               21                6              
32 32       9               31               15               16              
33 33       9               32               14                8              
34 34       9               24               24                4              
35 35       8               29               28                7              
36 36       9               27                5               11              
37 37       8               23               34               13              
38 38       9               33               26               12              
39 39      11               29               35                7              
40 40      12               33               21                2              
41 41       6               27               15                4              
42 42      11               29               22               14              
43 43       6               27               29               11              
44 44       7               25               19               10              
45 45       6               34               21                7              
46 46       9               23               15               15              
47 47       8               26               28                5              
48 48       6               21               24               11              
49 49      12               36               20                5              
50 50      10               36               17      

Самостоятельно сформулировать ограничения (максимальные или минимальные) на общую стоимость развертывания всей технической инфраструктуры КИС, возможные ограничения на суммарную стоимость отдельных типов станций ($C_1$, $C_2$, $C_3$, $C_4$), нижние ограничения на частоту процессора для $C_1, C_2$, другие приемлемые ограничения.

Построить регрессионную модель для оценки стоимости отдельной рабочей станции в зависимости от частоты процессора, объема оперативной памяти, объема винчестера.

Сперва решим задачу для нижнего уровня: отберём наиболее оптимальные компьютеры для каждого сегмента по соотношению цена-качество. У каждого сегмента свои критерии отбора, но везде цена минимизируется, остальные параметры максимизируются

## Находим зависимость стоимости от других параметров

In [ ]:
p_price<-lm(price~speed+hd+ram+screen, DataComputers)
p_price


Call:
lm(formula = price ~ speed + hd + ram + screen, data = DataComputers)

Coefficients:
(Intercept)        speed           hd          ram       screen  
    10.3331       5.2493      -0.5794      76.7454     105.5259  


## Находим средние значения

In [ ]:
mean_ram<-mean(DataComputers$ram)
mean_speed<-mean(DataComputers$speed)
mean_hd<-mean(DataComputers$hd)
mean_screen<-mean(DataComputers$screen)

mean_ram
mean_speed
mean_hd
mean_screen

[1] 8.286947

[1] 52.01102

[1] 416.6017

[1] 14.60872

## Задаем ограничения для C1-C4



In [ ]:
install.packages("hash", lib = "/usr/local/lib/R/site-library")
library("hash")

hash-2.2.6.2 provided by Decision Patterns




In [ ]:

restr_c1<-hash()
restr_c1[["price"]]<-50000
restr_c1[["min_n"]]<-3
restr_c1[["max_n"]]<-9
restr_c1[["hd"]]<-500
restr_c1[["ram"]]<-16
restr_c1[["screen"]]<-15
restr_c1[["multi"]]<-0
restr_c1[["speed"]]<-50
#C2
restr_c2<-hash()
restr_c2[["price"]]<-50000
restr_c2[["min_n"]]<-12
restr_c2[["max_n"]]<-32
restr_c2[["hd"]]<-500
restr_c2[["ram"]]<-24
restr_c2[["screen"]]<-15
restr_c2[["multi"]]<-0
restr_c2[["speed"]]<-30
#C3
restr_c3<-hash()
restr_c3[["price"]]<-40000
restr_c3[["min_n"]]<-5
restr_c3[["max_n"]]<-9
restr_c3[["hd"]]<-200
restr_c3[["ram"]]<-8
restr_c3[["screen"]]<-14
restr_c3[["multi"]]<-"yes"
restr_c3[["speed"]]<-25
#C4
restr_c4<-hash()
restr_c4[["price"]]<-20000
restr_c4[["min_n"]]<-9
restr_c4[["max_n"]]<-12
restr_c4[["hd"]]<-200
restr_c4[["ram"]]<-8
restr_c4[["screen"]]<-15
restr_c4[["multi"]]<-"yes"
restr_c4[["speed"]]<-25

## Задаем коэффициенты для частных критериев нижнего уровня

In [ ]:
w1<-c(0.8,0.3) #speed и ram в C1
w2<-c(0.3,0.7) # ram и hd в C2 
w3<-c(0.5,0.5) 
w4<-c(0.4,0.7)
h2<-c(0.25,0.25,0.25,0.25)
h2_goal<-w1[1]*h2[1]*mean_speed+w1[2]*h2[1]*mean_ram+ w2[1]*h2[2]*mean_ram+w2[2]*h2[2]*mean_hd+ w3[1]*h2[3]*mean_speed+w3[2]*h2[3]*mean_screen+w4[1]*h2[4]*mean_speed+w4[2]*h2[4]*mean_hd

## Решаем двухуровневую задачу оптимизации

In [ ]:
#В целевой функции H1 - общая стоимость системы:
#для C1 = 10.3331*N1+5.2493*N1*speed+76.7454*N1*ram-0.5794*N1*hd+105.5259*N1*screen
#для C2 = 10.3331*N2+5.2493*N2*speed+76.7454*N2*ram-0.5794*N2*hd+105.5259*N2*screen
#для C3 = 10.3331*N3+5.2493*N3*speed+76.7454*N3*ram-0.5794*N3*hd+105.5259*N3*screen
#для C4 = 10.3331*N4+5.2493*N4*speed+76.7454*N4*ram-0.5794*N4*hd+105.5259*N4*screen


In [ ]:
#x1=10.3331*N1;   x2= N1*speed,   x3= N1*hd,   x4= N1*ram,   x5= N1*screen,
#x6=10.3331*N2;   x7= N2*speed,   x8= N2*hd,   x9= N2*ram,   x10= N2*screen,
#x11=10.3331*N3;  x12= N3*speed,  x13= N3*hd,  x14= N4*ram,  x15= N3*screen,
#x16=10.3331*N4;  x17= N4*speed,  x18= N4*hd,  x19= N4*ram,  x20= N4*screen,

In [ ]:
# Учитываем условия работоспособонсти:
mean_hd<-c(max(mean_hd, restr_c1[["hd"]]),max(mean_hd, restr_c2[["hd"]]),
           max(mean_hd, restr_c3[["hd"]]),max(mean_hd, restr_c4[["hd"]]))

mean_ram<-c(max(mean_ram, restr_c1[["ram"]]),max(mean_ram, restr_c2[["ram"]]),
            max(mean_ram, restr_c3[["ram"]]),max(mean_ram, restr_c4[["ram"]]))

mean_speed<-c(max(mean_speed, restr_c1[["speed"]]),max(mean_speed, restr_c2[["speed"]]),
              max(mean_speed, restr_c3[["speed"]]),max(mean_speed, restr_c4[["speed"]]))

mean_screen<-c(max(mean_screen, restr_c1[["screen"]]),max(mean_screen, restr_c2[["screen"]]),
              max(mean_screen, restr_c3[["screen"]]),max(mean_screen, restr_c4[["screen"]]))


In [ ]:
install.packages("lpSolve")
library(lpSolve)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
objective.in<-c(1, p_price$coefficients[2], p_price$coefficients[3], p_price$coefficients[4],
                p_price$coefficients[5], 1, p_price$coefficients[2], p_price$coefficients[3],
                p_price$coefficients[4], p_price$coefficients[5], 1, p_price$coefficients[2],
                p_price$coefficients[3], p_price$coefficients[4], p_price$coefficients[5], 
                1, p_price$coefficients[2], p_price$coefficients[3], p_price$coefficients[4],
                p_price$coefficients[5])
const.mat<-matrix(c(0,w1[1]*h2[1],0,w1[2]*h2[1],0,0,0,w2[2]*h2[2],w2[1]*h2[2],0,0,w3[1]*h2[3],0,0,w3[2]*h2[3],0,w4[1]*h2[4],w4[2]*h2[4],0,0,
                    1,p_price$coefficients[2],p_price$coefficients[3],p_price$coefficients[4],p_price$coefficients[5],0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,1,p_price$coefficients[2],p_price$coefficients[3],p_price$coefficients[4],p_price$coefficients[5],0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,1,p_price$coefficients[2],p_price$coefficients[3],p_price$coefficients[4],p_price$coefficients[5],0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,p_price$coefficients[2],p_price$coefficients[3],p_price$coefficients[4],p_price$coefficients[5],
                    1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,  
                    0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,
                    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1), nrow=45, byrow=TRUE)
const.dir<-c(">=","<=","<=","<=", "<=",
             ">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",
             ">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",
             ">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",">=","<=",">=","<=")
const.rhs<-c(h2_goal,restr_c1[["price"]],restr_c2[["price"]],restr_c3[["price"]],restr_c4[["price"]],
             restr_c1[["min_n"]]*p_price$coefficients[1],restr_c1[["max_n"]]*p_price$coefficients[1],
             restr_c1[["min_n"]]*mean_speed[1],restr_c1[["max_n"]]*mean_speed[1],
             restr_c1[["min_n"]]*mean_hd[1],restr_c1[["max_n"]]*mean_hd[1],
             restr_c1[["min_n"]]*mean_ram[1],restr_c1[["max_n"]]*mean_ram[1],
             restr_c1[["min_n"]]*mean_screen[1],restr_c1[["max_n"]]*mean_screen[1],
             restr_c2[["min_n"]]*p_price$coefficients[1],restr_c2[["max_n"]]*p_price$coefficients[1],
             restr_c2[["min_n"]]*mean_speed[2],restr_c2[["max_n"]]*mean_speed[2],
             restr_c2[["min_n"]]*mean_hd[2],restr_c2[["max_n"]]*mean_hd[2],
             restr_c2[["min_n"]]*mean_ram[2],restr_c2[["max_n"]]*mean_ram[2],
             restr_c2[["min_n"]]*mean_screen[2],restr_c2[["max_n"]]*mean_screen[2],
             restr_c3[["min_n"]]*p_price$coefficients[1],restr_c3[["max_n"]]*p_price$coefficients[1],
             restr_c3[["min_n"]]*mean_speed[3],restr_c3[["max_n"]]*mean_speed[3],
             restr_c3[["min_n"]]*mean_hd[3],restr_c3[["max_n"]]*mean_hd[3],
             restr_c3[["min_n"]]*mean_ram[3],restr_c3[["max_n"]]*mean_ram[3],
             restr_c3[["min_n"]]*mean_screen[3],restr_c3[["max_n"]]*mean_screen[3],
             restr_c4[["min_n"]]*p_price$coefficients[1],restr_c4[["max_n"]]*p_price$coefficients[1],
             restr_c4[["min_n"]]*mean_speed[4],restr_c4[["max_n"]]*mean_speed[4],
             restr_c4[["min_n"]]*mean_hd[4],restr_c4[["max_n"]]*mean_hd[4],
             restr_c4[["min_n"]]*mean_ram[4],restr_c4[["max_n"]]*mean_ram[4],
             restr_c4[["min_n"]]*mean_screen[4],restr_c4[["max_n"]]*mean_screen[4])
res<-lp("min",objective.in, const.mat, const.dir, const.rhs)
res$constraints

[,1]     [,2]         [,3]         [,4]         [,5]        
                0.0000     1.000000     0.000000     0.000000     0.000000
                0.2000     5.249299     0.000000     0.000000     0.000000
                0.0000    -0.579360     0.000000     0.000000     0.000000
                0.0750    76.745450     0.000000     0.000000     0.000000
                0.0000   105.525919     0.000000     0.000000     0.000000
                0.0000     0.000000     1.000000     0.000000     0.000000
                0.0000     0.000000     5.249299     0.000000     0.000000
                0.1750     0.000000    -0.579360     0.000000     0.000000
                0.0750     0.000000    76.745450     0.000000     0.000000
                0.0000     0.000000   105.525919     0.000000     0.000000
                0.0000     0.000000     0.000000     1.000000     0.000000
                0.1250     0.000000     0.000000     5.249299     0.000000
                0.0000     0.000000     0.000000    -0.579360     0.000000
                0.0000     0.000000     0.000000    76.745450     0.000000
                0.1250     0.000000     0.000000   105.525919     0.000000
                0.0000     0.000000     0.000000     0.000000     1.000000
                0.1000     0.000000     0.000000     0.000000     5.249299
                0.1750     0.000000     0.000000     0.000000    -0.579360
                0.0000     0.000000     0.000000     0.000000    76.745450
                0.0000     0.000000     0.000000     0.000000   105.525919
const.dir.num   2.0000     1.000000     1.000000     1.000000     1.000000
const.rhs     170.9844 50000.000000 50000.000000 40000.000000 20000.000000
              [,6]     [,7]     [,8]     [,9]     [,10] [,11] [,12]    [,13]   
               1.00000  1.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   1.0000   1.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    1  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      1.00000   1.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
               0.00000  0.00000   0.0000   0.0000    0  ⋯      0.00000   0.0000
const.dir.num  2.00000  1.00000   2.0000   1.0000    2  ⋯      2.00000   1.0000
const.rhs     30.99932 92.99796 156.0331 468.0992 1500  ⋯     92.99796 123.9973
              [,14]    [,15]    [,16]    [,17]   [,18]    [,19]    [,20] [,21]
                0.0000   0.0000    0.000    0.00  0.00000  0.00000   0     0  
                0.0000   0.0000    0.000    0.00  0.00000  0.00000   0     0  
                0.0000   0.0000    0.000    0.00  0.00000  0.00000   0     0  
                0.0000   0.0000    0.000    0.00  0.00000  0.00000   0     0  
                0.0000   0.0000    0.000    0.00  0.00

## Проверим

In [ ]:
HH1<-function(p)
{ return(sum(c(1, p_price$coefficients[2], p_price$coefficients[3], p_price$coefficients[4],
               p_price$coefficients[5], 1, p_price$coefficients[2], p_price$coefficients[3],
               p_price$coefficients[4], p_price$coefficients[5], 1, p_price$coefficients[2],
               p_price$coefficients[3], p_price$coefficients[4], p_price$coefficients[5], 
               1, p_price$coefficients[2], p_price$coefficients[3], p_price$coefficients[4],
               p_price$coefficients[5])*p))
}
HH2<-function(p){return(sum(c(0,w1[1]*h2[1],0,w1[2]*h2[1],0,0,0,w2[2]*h2[2],w2[1]*h2[2],0,0,0,w3[2]*h2[3],w3[1]*h2[3],0,0,w4[1]*h2[4],w4[2]*h2[4],0,0)*p))}
HH1(res$solution)
HH2(res$solution)

[1] 71659.38

[1] 4251.936

## Проверяем условия роботоспособности

In [ ]:
apply_restr<-function(data, restr) 
{
  new_data<-data.frame()
  for (i in c(1:length(rownames(data))))
  {
    if (restr[["multi"]]!=0)
      if ((data[i,3]>=restr[["speed"]]) & (data[i,4]>=restr[["hd"]]) 
          & (data[i,6]>=restr[["screen"]]) & (data[i,5]>=restr[["ram"]])
          & (data[i,8]==restr[["multi"]]))
            new_data<-rbind(new_data,data[i,])
    if (restr[["multi"]]==0)
      if ((data[i,3]>=restr[["speed"]]) & (data[i,4]>=restr[["hd"]])
          & (data[i,6]>=restr[["screen"]]) & (data[i,5]>=restr[["ram"]]))
        new_data<-rbind(new_data,data[i,])
  }
  return(new_data)
}

new_computers_c1<-apply_restr(DataComputers, restr_c1)
new_computers_c2<-apply_restr(DataComputers, restr_c2)
new_computers_c3<-apply_restr(DataComputers, restr_c3)
new_computers_c4<-apply_restr(DataComputers, restr_c4)

## Нормализация

In [ ]:
normalize<-function(data)
{
  new_data<-data
  mmax<-max(data$price); mmin<-min(data$price)
  price<-(mmax-new_data$price)*100/(mmax-mmin)
  mmax<-max(data$speed); mmin<-min(data$speed)
  speed<-(new_data$speed-mmin)*100/(mmax-mmin)
  mmax<-max(data$hd); mmin<-min(data$hd)
  hd<-(new_data$hd-mmin)*100/(mmax-mmin)
  mmax<-max(data$ram); mmin<-min(data$ram)
  ram<-(new_data$ram-mmin)*100/(mmax-mmin)
  mmax<-max(data$screen); mmin<-min(data$screen)
  screen<-(new_data$screen-mmin)*100/(mmax-mmin)
  result<-data.frame(price,speed,hd,ram,screen)
  rownames(result)<-rownames(new_data)
  return(result)
}

norm_computers_c1<-normalize(new_computers_c1)
norm_computers_c2<-normalize(new_computers_c2)
norm_computers_c3<-normalize(new_computers_c3)
norm_computers_c4<-normalize(new_computers_c4)

## Доминация по Паретто

In [ ]:
DPareto<-function(X,Y){
  p<-TRUE; l<-FALSE; i<-1;
  while (p & (i<=length(X))){
    if (X[i]<Y[i]) p<-FALSE
    if (X[i]>Y[i]) l<-TRUE
    i<-i+1
  }
  if (!p | !l) return(FALSE)
  else return(TRUE)
}

## Ищем оптимальные решения по Паретто

In [ ]:
pareto_opt<-function(data)
{
  result<-c()
  for (i in c(1:length(rownames(data)))){
    p<-TRUE
    for (j in c(1:length(rownames(data)))){
      if (DPareto(data[j,],data[i,]))
        p<-FALSE}
    if (p) result<-c(result, rownames(data)[i])
  }
  return(result)  
}

In [ ]:
pareto_c1<-pareto_opt(norm_computers_c1)
pareto_c2<-pareto_opt(norm_computers_c2)
pareto_c3<-pareto_opt(norm_computers_c3)
pareto_c4<-pareto_opt(norm_computers_c4)

## Определяем идеальные точки

In [ ]:

N1<-res$solution[1]/p_price$coefficients[1]
N2<-res$solution[6]/p_price$coefficients[1]
N3<-res$solution[11]/p_price$coefficients[1]
N4<-res$solution[16]/p_price$coefficients[1]
ideal_c1<-c(res$solution[2]/N1,res$solution[3]/N1,res$solution[4]/N1,res$solution[5]/N1)
ideal_c2<-c(res$solution[7]/N2,res$solution[8]/N2,res$solution[9]/N2,res$solution[10]/N2)
ideal_c3<-c(res$solution[12]/N3,res$solution[13]/N3,res$solution[14]/N3,res$solution[15]/N3)
ideal_c4<-c(res$solution[17]/N4,res$solution[18]/N4,res$solution[19]/N4,res$solution[20]/N4)

## Находим расстояние между точками

In [ ]:

distance<-function(A, B) {return(sqrt(sum((A-B)^2)))}
for (i in c(1,2,3,4))
{
  if (i==1)
  {
    k<-pareto_c1[1]
    k_min<-distance(ideal_c1, new_computers_c1[k,c(3:6)])
    print("Для C1:")
    for (j in pareto_c1)
    {
      print(paste(j, ":", as.character(distance(ideal_c1, new_computers_c1[j,c(3:6)]))))
      if (distance(ideal_c1, new_computers_c1[j,c(3:6)])<k_min)
      {
        k_min<-distance(ideal_c1, new_computers_c1[j,c(3:6)])
        k<-j
      }     
    }
    cat("\n")
    print(paste("Лучший для C1: ",k," : ", k_min))
    cat("\n")
  }
  else
    if (i==2)
    {
      k<-pareto_c2[1]
      k_min<-distance(ideal_c2, new_computers_c2[k,c(3:6)])
      print("Для C2:")
      for (j in pareto_c2)
      {
        print(paste(j, ":", as.character(distance(ideal_c2, new_computers_c2[j,c(3:6)]))))
        if (distance(ideal_c2, new_computers_c2[j,c(3:6)])<k_min)
        {
          k_min<-distance(ideal_c2, new_computers_c2[j,c(3:6)])
          k<-j
        }  
      }
      cat("\n")
      print(paste("Лучший для C2: ",k," : ", k_min))
      cat("\n")
    }
    else
      if (i==3)
      {
        k<-pareto_c3[1]
        k_min<-distance(ideal_c3, new_computers_c3[k,c(3:6)])
        print("Для C3:")
        for (j in pareto_c3)
        {
          print(paste(j, ":", as.character(distance(ideal_c3, new_computers_c3[j,c(3:6)]))))
          if (distance(ideal_c3, new_computers_c3[j,c(3:6)])<k_min)
          {
            k_min<-distance(ideal_c3, new_computers_c3[j,c(3:6)])
            k<-j
          }  
        }
        cat("\n")
        print(paste("Лучший для C3: ",k," : ", k_min))
        cat("\n")
      }
      else
        if (i==4)
        {
          k<-pareto_c4[1]
          k_min<-distance(ideal_c4, new_computers_c4[k,c(3:6)])
          print("Для C4:")
          for (j in pareto_c4)
          {
            print(paste(j, ":", as.character(distance(ideal_c4, new_computers_c4[j,c(3:6)]))))
            if (distance(ideal_c4, new_computers_c4[j,c(3:6)])<k_min)
            {
              k_min<-distance(ideal_c4, new_computers_c4[j,c(3:6)])
              k<-j
            }  
          }
          cat("\n")
          print(paste("Лучший для C4: ",k," : ", k_min))
          cat("\n")
        }
}

[1] "Для C1:"
[1] "4847 : 500.072039028411"
[1] "5660 : 770.345697081362"
[1] "5882 : 244.750774882341"
[1] "5961 : 601.916058770246"
[1] "6045 : 500.199651585268"
[1] "6068 : 300.992513215477"
[1] "6083 : 300.325975310202"
[1] "6094 : 300.325975310202"
[1] "6118 : 300.992513215477"
[1] "6137 : 500.199651585268"
[1] "6147 : 300.992513215477"
[1] "6149 : 300.985868458586"
[1] "6158 : 502.297662552302"
[1] "6162 : 303.919301469167"
[1] "6166 : 500.199651585268"
[1] "6168 : 500.195653165863"
[1] "6169 : 303.813992116075"
[1] "6175 : 502.301644239302"
[1] "6181 : 300.325975310202"
[1] "6194 : 103.867670676537"
[1] "6200 : 112.0666846369"
[1] "6201 : 103.848413621824"
[1] "6203 : 303.919301469167"
[1] "6206 : 102.23351429949"
[1] "6213 : 303.925882092174"
[1] "6222 : 303.919301469167"
[1] "6231 : 303.925882092174"
[1] "6236 : 103.867670676537"
[1] "6238 : 111.206752517581"
[1] "6240 : 112.0666846369"
[1] "6244 : 101.290134988676"
[1] "6245 : 103.848413621824"
[1] "6252 : 102.23351429949"

[

Итого: <br>
- сегмент $C_1$: номер варианта 6244
- сегмент $C_2$: номер варианта 6149
- сегмент $C_3$: номер варианта 5822
- сегмент $C_4$: номер варианта 5619